controllers/networkfence_controller.go

```diff
@@ -39,17 +72,90 @@ type NetworkFenceReconciler struct {
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			logger.Info("NetworkFence resource not found")
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	// validate NetworkFence object so as its parameters are neither empty nor nil.
+	err = validateNetworkFenceSpec(nwFence)
+	if err != nil {
+		logger.Error(err, "failed to validate NetworkFence spec")
+
+		nwFence.Status.Result = csiaddonsv1alpha1.FencingOperationResultFailed
+		nwFence.Status.Message = fmt.Sprintf("Failed to validate Networkfence parameters: %v", util.GetErrorMessage(err))
+		statusErr := r.Client.Status().Update(ctx, nwFence)
+		if statusErr != nil {
+			logger.Error(statusErr, "Failed to update networkfence status")
+
+			return ctrl.Result{}, statusErr
+		}
+
+		// invalid parameter, do not requeue
+		return ctrl.Result{}, nil
+	}
 
-	// TODO(user): your logic here
+	logger = logger.WithValues("DriverName", nwFence.Spec.Driver, "CIDRs", nwFence.Spec.Cidrs)
+
+	conns := r.Connpool.GetByNodeID(nwFence.Spec.Driver, "")
+
+	// Iterate through the connections and find the one that matches the driver name
+	// provided in the NetworkFence spec; so that corresponding network fence and
+	// unfence operations can be performed.
+	for _, v := range conns {
+		for _, cap := range v.Capabilities {
+			// validate if NETWORK_FENCE capability is supported by the driver.
+			if cap.GetNetworkFence() == nil {
+				continue
+			}
+
+			// check if the networkfence object is getting deleted, if yes, then send
+			// UnfenceCluster request to the driver.
+			// After the UnfenceCluster request is processed, the finalizer will be removed
+			// so that the networkfence object can be deleted gracefully.
+			if !nwFence.GetDeletionTimestamp().IsZero() {
+				if util.ContainsInSlice(nwFence.GetFinalizers(), networkFenceFinalizer) {
+
+					err := r.unfenceClusterNetwork(ctx, nwFence, v.Client, logger)
+					if err != nil {
+						logger.Error(err, "failed to unfence cluster network")
+						return ctrl.Result{}, err
+					}
+				}
+				logger.Info("NetworkFence object is terminated, skipping reconciliation")
+				return ctrl.Result{}, nil
+			}
+
+			// add finalizer to the networkfence object if not already present.
+			if err := r.addFinalizerToNetworkFence(ctx, &logger, nwFence); err != nil {
+				logger.Error(err, "Failed to add NetworkFence finalizer")
+				return reconcile.Result{}, err
+			}
+
+			// validate of NETWORK_FENCE capability is enabled by the storage driver.
+			if cap.GetNetworkFence().GetType() != identity.Capability_NetworkFence_NETWORK_FENCE {
+				logger.Info("NetworkFence operation is not supported")
+				return ctrl.Result{}, nil
+			}
+
+			err := r.fenceClusterNetwork(ctx, nwFence, v.Client, logger)
+			if err != nil {
+				logger.Error(err, "failed to fence cluster network")
+				return ctrl.Result{}, err
+			}
+		}
+
+	}
```

Comment:
Update status, with successfully fenced state FencingOperationResultSucceeded?

Reply:
That doesn't update/patch the object; so got rid of it

Reply:
how will user know that the cluster is successfully fenced?

Reply:
Now re-wrote in such a way that status will be updated.